### PR TITLE
YC-586 - Restrict non public fillable forms to membre of administrative entity

### PR DESCRIPTION
**Pull request by monodo**
- [x] Check what happen when not form is available to user on administrative entity selection form
- [x] Adapt queryset to restrict visibility within administrative entity using group level permission "see_private_submissions"
- [x] Allow integrators to fill any form
- [x] Adapt tests

### CODE REVIEW

File: geocity/apps/submissions/views.py [hunk 1]
```diff
@@ -1132,6 +1132,8 @@ def submission_select_administrative_entity(request, submission_id=None):
         entity_tags=entity_tags,
     )
 
+    print(entities)
```
Comment (rbovard) [inline geocity/apps/submissions/views.py:1135]: Remove that

**Comment by rbovard** [comment]
Something is not working.

An user in a group with `see_private_submissions` permission should see **ONLY** the private forms of the admin entity of the group he's belong to. Currently, he can see **ALL** private forms if he's in a group with `see_private_submissions` permission.

**Review by rbovard — CHANGES_REQUESTED** [review]
Something is not working.

An user in a group with `see_private_submissions` permission should see **ONLY** the private forms of the admin entity of the group he's belong to. Currently, he can see **ALL** private forms if he's in a group with `see_private_submissions` permission.

*Edit: when I said "see private form", it's public + private forms of course*

**Review by rbovard — COMMENTED** [review]
It's working fine, thanks!

File: geocity/apps/submissions/forms.py [hunk 2]
```diff
@@ -181,8 +181,37 @@ def __init__(self, instance, form_categories=None, *args, **kwargs):
...
+                he can fill private forms for those administrative entities
+                if granted permission 'view_private_submission'"""
+                forms_filter &= Q(
+                    administrative_entities__in=user_administrative_entities
+                ) | Q(is_public=True)
+            elif (
+                not user_can_view_private_submission or not user_administrative_entities
+            ):
+                """Untrusted users or user not granted with view_private_submission can only fill public forms"""
+                forms_filter &= Q(is_public=True)
```
Comment (AlexandreJunod) [inline geocity/apps/submissions/forms.py:207]: use `get_administrative_entities_with_forms`

File: geocity/apps/submissions/forms.py [hunk 3]
```diff
@@ -181,8 +181,37 @@ def __init__(self, instance, form_categories=None, *args, **kwargs):
...
+
+        if user_administrative_entities and user_can_view_private_submission:
+            """User is trusted and associated to administrative entities,
+            he can fill private forms for this administrative entity"""
+            forms_filter &= Q(
+                administrative_entities__in=user_administrative_entities
+            ) | Q(is_public=True)
```
Comment (AlexandreJunod) [inline geocity/apps/submissions/forms.py:214]: Same condition as line 196 to 202

**Review by AlexandreJunod — CHANGES_REQUESTED** [review]

**Review by AlexandreJunod — APPROVED** [review]

File: geocity/apps/submissions/forms.py [hunk 4]
```diff
@@ -202,8 +202,30 @@ def __init__(self, instance, form_categories=None, *args, **kwargs):
...
+                he can fill private forms for those administrative entities
+                if granted permission 'view_private_submission'"""
+                forms_filter &= Q(
+                    administrative_entities__in=user_administrative_entities
+                ) | Q(is_public=True)
+            elif (
+                not user_can_view_private_submission or not user_administrative_entities
+            ):
+                """Untrusted users or user not granted with view_private_submission can only fill public forms"""
+                forms_filter &= Q(is_public=True)
```
Comment (monodo) [inline geocity/apps/submissions/forms.py:228]: Couldn't we avoid to duplicate logic that is already here: https://github.com/yverdon/geocity/pull/676/files#diff-401ab1dedecd4ec91024d12d0d554a535a6edde585f43caffb7147368fec1ebbR105 ?

**Comment by rbovard** [comment]
@AlexandreJunod Thanks a lot, if it's ok for you I'll test after you adress @monodo review.

**Comment by AlexandreJunod** [comment]
Add lock https://fontawesome.com/icons/lock?s=regular&f=classic on the list of private forms so we can distinguish them

File: geocity/apps/accounts/models.py [hunk 5]
```diff
@@ -21,10 +21,18 @@
 
 from .fields import AdministrativeEntityFileField
 
-# public types: for public/restricted features
+# Controls who can fill each Form
```
[the remark this answers is not in the document]
Comment (AlexandreJunod) [reply geocity/apps/accounts/models.py:24]: anonymous forms can be filled by anyone

**Comment by AlexandreJunod** [comment]
@rbovard monodo review is done, tests should normally pass, you can test

**Review by rbovard — CHANGES_REQUESTED** [review]
The visiblity of private/public form is working well but I noticed a regression.

When there is only one form in an entity, the user should be redirected to the step "Details" and not "Objets" (it's the case in current `develop`)

**Comment by monodo** [comment]
> re is only one form in an entity, the user should be r



> The visiblity of private/public form is working well but I noticed a regression.
> 
> When there is only one form in an entity, the user should be redirected to the step "Details" and not "Objets" (it's the case in current `develop`)

Cannot reproduce on local instance 🤔

**Comment by rbovard** [comment]
image

**Comment by monodo** [comment]
> image

Looks like its related to specific config. I suggest to open a specific issue.

**Review by rbovard — APPROVED** [review]
> Looks like its related to specific config. I suggest to open a specific issue.

https://jira.liip.ch/browse/YC-1049

**Pull request merged**